### PR TITLE
Define schema for Kptfile v1alpha2

### PR DESCRIPTION
**Pull request by frankfarzan**
Defines the core schema.

Some additive changes not included for now as they're being finalized (will be additive change in separate PRs):

- Setters as a top-level field in `pipeline`.
- Filters in `pipeline`.

### CODE REVIEW

File: pkg/api/kptfile/v1alpha2/types.go
Comment (Shell32-Natsu) [inline pkg/api/kptfile/v1alpha2/types.go:135]: Can we use the `Pipeline` and `Function` defined in `internal/pipeline`? Otherwise we need to move a lot of codes here.

File: pkg/api/kptfile/v1alpha2/types.go
Comment (mortent) [inline pkg/api/kptfile/v1alpha2/types.go:85]: I'm not sure if we need to include this in the schema. Instead we can make the schema extensible and allow users to add any information they need.

File: pkg/api/kptfile/v1alpha2/types.go
Comment (mortent) [inline pkg/api/kptfile/v1alpha2/types.go:116]: We either need a `Type` field here, or `Upstream` needs to be a pointer. We don't want to have to look at fields within the `Upstream` struct.

File: pkg/api/kptfile/v1alpha2/types.go
Comment (mortent) [inline pkg/api/kptfile/v1alpha2/types.go:80]: What does it mean to have an `UpdateStrategy` as part of `Upstream`? This means that the definition of the subpackage in the parent Kptfile and the upstream information could have different update strategies. It seems like just keeping it in the subpackage section of the parent Kptfile would be sufficient.

File: pkg/api/kptfile/v1alpha2/types.go
Comment (mortent) [reply pkg/api/kptfile/v1alpha2/types.go:135]: We are looking to move away from unmarshalling the Kptfile into structs, so I don't think we need to worry about it. I think this structure is mostly about communicating the structure.

File: pkg/api/kptfile/v1alpha2/types.go
Comment (mortent) [inline pkg/api/kptfile/v1alpha2/types.go:26]: We might want to use the `schema.GroupVersion` type from apimachinery here. We should also include the kind.

File: pkg/api/kptfile/v1alpha2/types.go
Comment (phanimarupaka) [reply pkg/api/kptfile/v1alpha2/types.go:85]: Yes. I am making changes to treat Kptfile as a KRM resource.

File: pkg/api/kptfile/v1alpha2/types.go
Comment (phanimarupaka) [inline pkg/api/kptfile/v1alpha2/types.go:109]: Can this also include its own Pipeline definition? So that the transformations specific to the package are specified here. With this structure, if users want to perform operations specific to a subpackage, they should include filters in the main pipeline. By including the Pipeline definition here, it is easy to read.

File: pkg/api/kptfile/v1alpha2/types.go
Comment (phanimarupaka) [inline pkg/api/kptfile/v1alpha2/types.go:156]: Shouldn't this be []Mutator which in turn contains Function definition? Where can we specify filters here ? And additional metadata which is specific to scoping?

File: pkg/api/kptfile/v1alpha2/types.go
Comment (frankfarzan) [reply pkg/api/kptfile/v1alpha2/types.go:80]: Just considering the single-package UX for sake of this discussion: Do we want the update strategy of the package to be declared?

1. kpt pkg get foo
2. kpt pkg update

Should step 2 always specify the strategy as a flag? Or should it be recorder in the Kptfile?

Back to the subpackage UX, isn't this a general problem of having Upstream declared in the parent package (`.subpackage.upstream`) and in the subpackage (`.upstream`). You can edit other fields such as Git repo, directory, etc. which needs to be reconciled.

File: pkg/api/kptfile/v1alpha2/types.go
Comment (frankfarzan) [reply pkg/api/kptfile/v1alpha2/types.go:85]: I'm not sure I agree. We can have certain common metadata declared in the schema and also enable extension through `tags` field. Many package managers such as NPM work this way.

File: pkg/api/kptfile/v1alpha2/types.go
Comment (frankfarzan) [reply pkg/api/kptfile/v1alpha2/types.go:116]: Yep, making this a pointer makes sense. In fact most fields in Kptfile should be pointers as they're optional.

File: pkg/api/kptfile/v1alpha2/types.go
Comment (frankfarzan) [reply pkg/api/kptfile/v1alpha2/types.go:109]: I'm not sure I parse this. What pattern are you thinking of? For references, we currently having Fork/Rebase and Sub-package overlay pattern, you might be thinking of different pattern alltogether.

File: pkg/api/kptfile/v1alpha2/types.go
Comment (frankfarzan) [reply pkg/api/kptfile/v1alpha2/types.go:156]: Bunch of different questions here:

"Shouldn't this be []Mutator which in turn contains Function definition?"

Elaborate? Example Pipeline: go/kpt-pipeline-declaration

"Where can we specify filters here ?"

Per PR description, this still need to be finalized and added to the schema. But it will be an additive change.

"And additional metadata which is specific to scoping?"

Elaborate? "scoping" is generally an overloaded word.

File: pkg/api/kptfile/v1alpha2/types.go
Comment (frankfarzan) [reply pkg/api/kptfile/v1alpha2/types.go:135]: The schema should be used in code. Discussion of how it's marshalled/unmarshalled from YAML is orthogonal to this. In addition, Kptfile needs to eventually have a CRD which can be generated from this.

@Shell32-Natsu This is mostly the same as the struct in internal/pipeline, so refactoring the pipeline code to use this should follow easily.

File: pkg/api/kptfile/v1alpha2/types.go
Comment (frankfarzan) [reply pkg/api/kptfile/v1alpha2/types.go:26]: Yes, makes sense.

File: pkg/api/kptfile/v1alpha2/types.go
Comment (phanimarupaka) [inline pkg/api/kptfile/v1alpha2/types.go:30]: How about SetterDefinitions?

Comment (frankfarzan) [reply pkg/api/kptfile/v1alpha2/types.go:30]: Per the PR description, this is still being finalized and will be added in the future.

File: pkg/api/kptfile/v1alpha2/types.go
Comment (mortent) [reply pkg/api/kptfile/v1alpha2/types.go:80]: So I think there is a couple of things here. 

I agree that being able to specify an update strategy on the root package declaratively in the Kptfile would be useful. It is a bit awkward the way it is now, where it needs to be specified as a flag on every update. We should see if there is a way to address that. It seems like it would require specifying in the package Kptfile, which seems less ideal for subpackages.

Having the git information in both the parent Kptfile and the package Kptfile serves two different purposes. The information in the package Kptfile tells us origin of the current content of the package, which we use for the 3-way merge. The information in the parent Kptfile is where the user specifies the desired version, i.e. the upstream in our 3-way merge. So we actually need both pieces of information, but it not clear to me what it means to have the update strategy in two places.

I think we want to use separate types for subpackage and upstream. Not only to avoid having update strategy in two places, but also to let us keep separate information about the upstream in the two places. Currently we share the `Git` struct between upstream and subpackages, which is a bit unfortunate. The git information specified in subpackages should only require `repo`, `directory` and `ref`, while the information in upstream also needs the `commit` field. We currently allow the `commit` field inside subpackage spec, which we shouldn't.

File: pkg/api/kptfile/v1alpha2/types.go
Comment (droot) [reply pkg/api/kptfile/v1alpha2/types.go:85]: Extension are definitely needed but making the pkg info first class sends a message about the intent to the kpt pkg authors about including this info. Also, we should make a point to use these fields in our examples pkgs and also surface in CLI experience when pkgs has these.  

on that note, suggestion about using `info` field in the kptfile to make it more user-friendly instead of `packageMeta`

File: pkg/api/kptfile/v1alpha2/types.go
Comment (frankfarzan) [reply pkg/api/kptfile/v1alpha2/types.go:80]: I think we're on the same page for most part. This is what I refer to as "reconciliation". Kpt reconciles the following:

1. Last fetched upstream: `kptfile.upstream`
2. Desired upstream: `kptfile.subpackage.upstream`

Keep in mind, we want to design the schema so it's possible to add other version control systems in the future without breaking backwards compatibility. `kptfile.upstream` has Git in a "one of" which is what we want.

I see couple of options:
1. Use different types for 1 and 2.
2. Reuse common type `Upstream` for both 1 and 2. 

"The git information specified in subpackages should only require repo, directory and ref, while the information in upstream also needs the commit field. We currently allow the commit field inside subpackage spec, which we shouldn't."

It should be fine to specify an exact commit id for the desired origin. All Git-based tooling supports this (Golang, Config Sync, etc.). Providing humna-readable refs should be optional not mandatory. Actually, I realize that `ref` is kind of ambiguous (under-documented here). Does it hold Git tags or branches?

File: pkg/api/kptfile/v1alpha2/types.go
Comment (droot) [reply pkg/api/kptfile/v1alpha2/types.go:135]: I think @Shell32-Natsu point is about code organization, are we moving all the types definition (atleast referred by Kptfile) in one place under `v1alpha2` or just having root type here that can refer to types defined in other packages. There is some code that is written in OOP style where methods are implemented on types for ex. `Function` and `Pipeline`, so it impacts that.

Having types in one place here might be simpler for the main types that Kptfile directly uses.  @Shell32-Natsu pipeline pkg's code will use the types defined here.

**Review by droot — COMMENTED** [review]
I have minor comments. Overall, looks good to me.

File: pkg/api/kptfile/v1alpha2/types.go
Comment (mortent) [reply pkg/api/kptfile/v1alpha2/types.go:80]: I think separating the two types makes sense and gives us the most flexibility going forward. In particular if we want to add support for other storage systems where the information needed for tracking "last fetched upstream" vs "desired upstream" might differ significantly. 

For the `Git` type, the `ref` field holds a git reference, which could be branch, tag or commit. This is the field used by users to specify desired upstream ref. The `commit` field on the other hand, is used to hold a reference to origin for the purpose of keeping track of the last fetched upstream. So even is the user specifies a tag in the `ref` field, kpt will find the commit sha and use that to track origin. We should consider the `commit` field as an implementation detail that is managed by kpt, while `ref` is the api for users (regardless if they want to specify branch, tag or short/long commit sha).

File: pkg/api/kptfile/v1alpha2/types.go
Comment (frankfarzan) [reply pkg/api/kptfile/v1alpha2/types.go:135]: Yes, we want to have all the types in Kptfile be self-contained in this package. The question is really about method receivers on these types. 

This is actually a much-asked feature require in Go as other language provide type extension:

https://github.com/golang/go/issues/21401

Choices are:

1. Use functions instead of method receivers
2. Put method receivers in this package
3. Type alias or embed these types. See https://stackoverflow.com/a/28800807

If the method receivers are generic and reusable across many features, 2 seems reasonable. If on the other hand, they're specific to one feature (e.g. Pipeline validation), then 1 and 3 are reasonable.

File: pkg/api/kptfile/v1alpha2/types.go
Comment (phanimarupaka) [reply pkg/api/kptfile/v1alpha2/types.go:156]: scoping: Where can a user specify filters to target only specific packages ? 

this still need to be finalized and added to the schema => got it.

File: pkg/api/kptfile/v1alpha2/types.go
Comment (phanimarupaka) [reply pkg/api/kptfile/v1alpha2/types.go:109]: you might be thinking of different pattern all together => This might not be the right place to discuss that. We can sync offline.

File: pkg/api/kptfile/v1alpha2/types.go
Comment (frankfarzan) [reply pkg/api/kptfile/v1alpha2/types.go:26]: Actually, isn't yaml.ResourceMeta embedded in the Kptfile the same schema?

File: pkg/api/kptfile/v1alpha2/types.go
Comment (frankfarzan) [reply pkg/api/kptfile/v1alpha2/types.go]: Renaming this field to `info` SGTM.

File: pkg/api/kptfile/v1alpha2/types.go
Comment (frankfarzan) [reply pkg/api/kptfile/v1alpha2/types.go:80]: Create two types: Upstream and UpstreamLock.

Comment (frankfarzan) [reply pkg/api/kptfile/v1alpha2/types.go:80]: @mortent This formulation makes it obvious that we don't have an `Upstream` for the top-level package; only a `UpstreamLock` (We were discussing this in the context of recording update strategy before).

File: pkg/api/kptfile/v1alpha2/types.go
Comment (martinmaly) [inline pkg/api/kptfile/v1alpha2/types.go:104]: Is this used for tool bookkeeping to track what the `ref` resolved into? Or can it be used as input?
Presumably `ref` is the input, and if the `ref` is a SHA, the Commit will inherit the same value?

File: pkg/api/kptfile/v1alpha2/types.go
Comment (martinmaly) [inline pkg/api/kptfile/v1alpha2/types.go:114]: How is the value for this field determined? Is it meaningful, i.e. wrt maintainers often being tracked using other means, such as GitHub roles? How would packages with multiple maintainers be described?

File: pkg/api/kptfile/v1alpha2/types.go
Comment (martinmaly) [inline pkg/api/kptfile/v1alpha2/types.go:117]: Is this the license inlined?  Apache license, for example, is rather large (https://www.apache.org/licenses/LICENSE-2.0.txt)

Or, is this a path to the license file included in the package? I'd recommend using a path.

File: pkg/api/kptfile/v1alpha2/types.go
Comment (martinmaly) [inline pkg/api/kptfile/v1alpha2/types.go:120]: Do we recommend semver (for example) or leave this undefined or without recommendation/guidance?

File: pkg/api/kptfile/v1alpha2/types.go
Comment (martinmaly) [inline pkg/api/kptfile/v1alpha2/types.go:123]: How is this related to the resource metadata, especially its labels and annotations?

File: pkg/api/kptfile/v1alpha2/types.go
Comment (martinmaly) [inline pkg/api/kptfile/v1alpha2/types.go:126]: Unless there is strong reason to base this on `man`, I'd recommend calling this `documentation` or `doc`.
the word `man` is also ambiguous, depending on context (i.e. person)

File: pkg/api/kptfile/v1alpha2/types.go
Comment (martinmaly) [inline pkg/api/kptfile/v1alpha2/types.go:193]: recommend s/K8S/KRM/

(note also different casing used throughout the file)

File: pkg/api/kptfile/v1alpha2/types.go
Comment (droot) [reply pkg/api/kptfile/v1alpha2/types.go:74]: Having an explicit field can help the API consumers. This post from OpenAPI spec --> https://swagger.io/docs/specification/data-models/inheritance-and-polymorphism/ talks about this. Note the `descriminator` field maps to `Type` in this case.

>One possible example are code generation tools: they can use discriminator to generate program statements that typecast request data to appropriate object type based on the discriminator property value.

Comment (mortent) [reply pkg/api/kptfile/v1alpha2/types.go:74]: Using a discriminator field is a pattern pretty commonly used in the built-in kubernetes types.

File: pkg/api/kptfile/v1alpha2/types.go
Comment (mortent) [reply pkg/api/kptfile/v1alpha2/types.go:80]: Yeah, I think the two different types makes sense. I think we should also create two separate types for the `Git` field in `Upstream` and `UpstreamLock`. The `Commit` field should only be available in `UpstreamLock`. The question around the use of this field on this PR suggests it is confusing.

File: pkg/api/kptfile/v1alpha2/types.go
Comment (mortent) [reply pkg/api/kptfile/v1alpha2/types.go:104]: This is only used for bookkeeping to track the upstream git SHA. It will always be set to a specific git SHA no matter if the provided `ref` is a branch, tag or long/short git commit.

File: pkg/api/kptfile/v1alpha2/types.go
Comment (mortent) [reply pkg/api/kptfile/v1alpha2/types.go:111]: I think these comments points to why we might want to leave this undefined (at least for now) and allow users to extend the schema with their own metadata. There are lots of different information users can add to a package and it is not clear to me why these fields are the right ones. And if we want to let users extend it anyway, why not let them just define everything. If any of this information will be used by kpt directly, then I agree it should be in the schema.

File: pkg/api/kptfile/v1alpha2/types.go
Comment (frankfarzan) [reply pkg/api/kptfile/v1alpha2/types.go:74]: I'm inclined to keep the discriminator field as is.

File: pkg/api/kptfile/v1alpha2/types.go
Comment (mikebz) [inline pkg/api/kptfile/v1alpha2/types.go:34]: seems like we had Upstream and now this is UpstreamLock.  Does it have the meaning of a lock?  Lock in NPM says that unless I update the refresh grabs the same versions.  Is that what we are doing and is that why we decided on a name change?

File: pkg/api/kptfile/v1alpha2/types.go
Comment (mikebz) [inline pkg/api/kptfile/v1alpha2/types.go:141]: It seems like we have two different definitions for subpackage and root package definition.  Upstream vs UpstreamLock.  Is the update strategy the difference maker here?  Why have two different ways and names to describe upstream info?

File: pkg/api/kptfile/v1alpha2/types.go
Comment (frankfarzan) [reply pkg/api/kptfile/v1alpha2/types.go:34]: There's a discussion below, but effectively yes. `UpstreamLock` is product of kpt resolving the user-specified `Upstream`.

File: pkg/api/kptfile/v1alpha2/types.go
Comment (frankfarzan) [reply pkg/api/kptfile/v1alpha2/types.go]: Changed this to be a list. Per the comment on this type, all fields in `PackageInfo` are informational and not consumed by any functionality in kpt.

File: pkg/api/kptfile/v1alpha2/types.go
Comment (frankfarzan) [reply pkg/api/kptfile/v1alpha2/types.go:137]: Split this into two: A SPDX identifier string and a path to a LICENSE file. We shouldn't be inlining the content of the license here.

File: pkg/api/kptfile/v1alpha2/types.go
Comment (frankfarzan) [reply pkg/api/kptfile/v1alpha2/types.go]: Renamed to `keywords` and added a comment to `PackageInfo` clarify this.

File: pkg/api/kptfile/v1alpha2/types.go
Comment (frankfarzan) [reply pkg/api/kptfile/v1alpha2/types.go]: I removed this field as it would create confusion with the concept of version that's used by kpt machinary.

File: pkg/api/kptfile/v1alpha2/types.go
Comment (frankfarzan) [reply pkg/api/kptfile/v1alpha2/types.go:223]: I'll follow up with @seans3 to expand the comments here.

File: pkg/api/kptfile/v1alpha2/types.go
Comment (frankfarzan) [reply pkg/api/kptfile/v1alpha2/types.go:80]: Split `Git` and `GitLock` for maximum decoupling/flexibility.

File: pkg/api/kptfile/v1alpha2/types.go
Comment (frankfarzan) [reply pkg/api/kptfile/v1alpha2/types.go]: Renamed to Site. As discussed offline, we need to define the schema for Kptfile. There are extension points via `metadata.annotations|labels` and `packageInfo.keywords` for humans and automation to add arbitrary metadata. There are also some common metadata that every package manager needs. We want to define that here instead of having each user of kpt independently create a convention for it.

File: pkg/api/kptfile/v1alpha2/types.go
Comment (frankfarzan) [reply pkg/api/kptfile/v1alpha2/types.go:161]: `Upstream` is specified, `UpstreamLock` is resolved by kpt. There's an asymmetry that there's no `Upstream` for a top-level package, it's purely CLI-driven.

Comment (mikebz) [reply pkg/api/kptfile/v1alpha2/types.go:161]: Questions:
Not sure I understand what UpstreamLock is resolved by kpt mean?
Why is it called Lock?  Is it locking something?

File: pkg/api/kptfile/v1alpha2/types.go
Comment (droot) [reply pkg/api/kptfile/v1alpha2/types.go:165]: closing the loop here: @Shell32-Natsu and me had an offline chat about this. There are very few methods defined on Pipeline and Function and turning them in to functions won't be a big deal. So all set.

File: pkg/api/kptfile/v1alpha2/types.go
Comment (seans3) [reply pkg/api/kptfile/v1alpha2/types.go:223]: @martinmaly You are correct--the first three `Inventory` values (Namespace, Name, and InventoryID) are filled in by `kpt live init`. These three values are needed locally 1) to find the inventory object in the cluster, or 2) to first create the inventory object if it does not yet exist in the cluster. Currently, `kpt live init` must first be run before any `kpt live apply`, and a **package not initialized error** will be thrown if this is not the case.

The namespace is calculated from the namespace of the resources in the package, or the `--namespace` flag during `init`, or a `namespace` set in the `kubeconfig` file. The name is currently `inventory-XXXXX` where XXXXX is a random number to ensure no name collision. We will probably change the name to be more meaningful by starting with the name of the directory. The `InventoryID` is a UUID which is added as a label to the cluster resource because we retrieve inventory objects by label (there used to the possibility of multiple inventory objects, and we also may need that capability for nested packages). I'm pretty sure the other two `Inventory` fields (`Labels` and `Annotations`) are added after `init` during the apply lifecycle.

**Review by droot — APPROVED** [review]
/lgtm

It will be good if we can get this merged (have some dependencies blocked on it) and address any comments in the follow up PR.

**Review by mortent — APPROVED** [review]

File: pkg/api/kptfile/v1alpha2/types.go
Comment (frankfarzan) [reply pkg/api/kptfile/v1alpha2/types.go:161]: So the user's intent is given by `Upstream`. kpt consumes `Upstream` and manages `UpstreamLock` with the commit SHA-1 which fully "locks" the the Git commit that was fetched. The only time `UpstreamLock` changes is when:
a) User updates `Upstream` for a sub-package OR
b) User invokes `kpt cfg update` for the top-level package.

File: pkg/api/kptfile/v1alpha2/types.go
Comment (frankfarzan) [reply pkg/api/kptfile/v1alpha2/types.go:223]: @seans3 Do you mind following up and adding a comment on each field here, and specifying whether it's user-specified or kpt managed? This is a public schema for Kpt, so we want each exported field to be well-documented.

**Pull request merged**
